Dart/src/com/jetbrains/lang/dart/ide/errorTreeView/DartProblem.java

```diff
@@ -167,10 +169,16 @@ private void ensureInitialized() {
   }
 
   public static @NotNull @Nls String generateTooltipText(@NotNull @Nls String message,
+                                                         @NotNull List<DartServerData.DartDiagnosticMessage> contextMessages,
                                                          @Nullable @Nls String correction,
                                                          @Nullable @NonNls String url) {
-
     HtmlBuilder htmlBuilder = new HtmlBuilder().append(message);
+    if (!contextMessages.isEmpty()) {
+      for (DartServerData.DartDiagnosticMessage contextMessage : contextMessages) {
+        htmlBuilder.append(HtmlChunk.br()).append(HtmlChunk.br()).append(contextMessage.getMessage());
```

Comment:
I think this is going to be turned into a hyperlink, but for the text of the link, instead of `(main.dart)`, could we do `main.dart:111`. I.e., file:line.

Reply:
I had that thought as well, but I am not sure if it is possible/pragmatic with the IJ APIs and what we have sent with the analysis server.  The `DiagnosticMessage`s don't include the line number, but do contain the `offset` and `length`.  For open files where the PSIs have been created, i.e. we have a `Document` object, we could use the `Document.getLineNumber(int offset)` method, but this could lead to lags in generating the hover tooltips.

I figured having `file.dart:<offset>` would be misleading.

@alexander-doroshko Is there a way to quickly convert an offset to line number without the PSI construction?

Reply:
>  Is there a way to quickly convert an offset to line number without the PSI construction?

Either via Document, or via FileOffsetsManager. The latter is the right choice if the offset comes from the server (it contains the required data but may lack public API for this task). Both don't parse files (don't create PSI) but both do load file contents (at least once). I think this is fast enough.

Reply:
Thanks Alex.  I updated the PR to use the `Document` API, the numbers out of the `FileOffsetsManager` were all offsets, not line numbers.  LMK if there are any changes you want before we land it.

Reply:
`Document` API assumes all line breaks are single-character. Offsets given by server count real line break characters, as they are on disk. So, Document API may give incorrect results for CR-LF line endings.

That's why we wrote `FileOffsetsManager` for Dart. 

Sorry, I wasn't clear. I mean, `FileOffsetsManager` has enough data (in its fields) to calculate `offset-> line number` correctly. But you are right, it doesn't have such public method yet. If we want to show line number correctly we need to add it.

Reply:
https://github.com/JetBrains/intellij-community/pull/1516

Reply:
Thanks! I'll merge both tomorrow.

Reply:
@jwren I just realized that we have `DiagnosticMessage.getLocation().getStartLine()` method. And we already se it in `DartProblemsViewPanel.addDiagnosticMessageActions()`. So we don't need any manual conversion at all! =)